### PR TITLE
Properly decode UTF-8 characters comming in from serial one byte at a time.

### LINKED ISSUE #2430: Input Method to compose complex characters
There are 2 bugs to use Korean characters:
1. If any ASCII character such as white space, comma, dot, or number is typed in during composing the Korean character, the currently composed Korean character is disappeared as like deleted in the screen of the Edit and the Serial Monitor windows. 
2. Korean characters are commonly consisted of 3 components: 1) beginning consonant, 2) vowel, and 3) final consonant. Sometimes the final consonant of the previous composed character becomes the beginning consonant of the following composed character. If the previous final consonant becomes the beginning consonant of the next composed character, then the previous composed Korean character should be redrawed with the final completely composed Korean character, but it is still displayed with the final consonant which becomes the beginning consonant of the next Korean character.
- Same in 1.5.x

### PR DESCRIPTION
**Pull request by aknrdureegaesr**
This fixes #2430.

### CODE REVIEW

**Comment by facchinm** [comment]
I can see some performance penalty here with sketches writing a lot of stuff.
@PaulStoffregen, what do you think about it?

**Comment by aknrdureegaesr** [comment]
My gut feeling is that the raw throughput of any such code (both mine and the previous) is _much_ higher than any serial port or any UI work can ever be.

If that helps any, I offer to provide performance tests to prove that point.

File: arduino-core/src/processing/app/Serial.java [hunk 1]
```diff
@@ -153,10 +167,24 @@ public synchronized void serialEvent(SerialPortEvent serialEvent) {
...
-          char[] chars = msg.toCharArray();
-          message(chars, chars.length);
+        int next = 0;
+        while(next < buf.length) {
+          while(next < buf.length && outToMessage.hasRemaining()) {
+            int spaceInIn = inFromSerial.remaining();
+            int copyNow = buf.length - next < spaceInIn ? buf.length - next : spaceInIn;
+            inFromSerial.put(buf, next, copyNow);
+            next += copyNow;
+            inFromSerial.flip();
```
Comment (cmaglie) [inline arduino-core/src/processing/app/Serial.java]: Why `flip()` is called?

Comment (aknrdureegaesr) [inline arduino-core/src/processing/app/Serial.java]: The buffer needs to be switched from being written to to being read from, which is what `flip` does.

Both `clear` and `compact` switch the buffer from having being read to being ready to accept writes.

**Comment by PaulStoffregen** [comment]
Ok, compiled and running it now.  Seems to be ok CPU-wise.  Tested only on Linux 64 bit running on a fast desktop machine.

**Comment by facchinm** [comment]
@ArduinoBot build this please.

**Comment by facchinm** [comment]
@ArduinoBot build this please.

**Comment by aknrdureegaesr** [comment]
What's the usual process with pull requests here in the Arduino repo? I thought, the obvious, "push a pull request and wait a few days".

But there has been nothing here for more than a week now. For that week, I've been waiting for the maintainers to move. I hope they are not simultaneously waiting for me.

Is anything missing I'm expected to provide? If so, I'm not aware of it. Please tell me.

Or what's been hindering this to be merged, if I may ask?

**Comment by matthijskooijman** [comment]
You have the process correct, but it typically takes more than a week for a PR to be thoroughly reviewed and tested, so it can be merged. From the comments, this PR is now waiting for @cmaglie or @facchinm to find some time to have another look, so I don't think any action is required on your side.

**Comment by cmaglie** [comment]
@aknrdureegaesr 
thanks for your contribution!

**Pull request merged**